### ISSUE TITLE
Composite index is very slow | Performance degradation

### ISSUE THREAD

**Issue by seeden** [orientechnologies/orientdb#5577]
Hi I have simple table with two columns: **locale**(String) and **created**(DateTime or Integer) and 60 000 reacords (OrientDB 2.0.8 and 2.1.6)
I have created composite nonunique index Quiz.created_locale.
There is only few locale variants like (en, it, hu) and each record has timestamp in column named created.

When I wrote:
```sql
select * from quiz where locale = 'hu' order by created desc limit 20
```

I can see:
**involvedIndexes**: Quiz.created_locale
**documentReads**: 41 603
**fullySortedByIndex**: true,
**indexIsUsedInOrderBy**: true,
**elapsed**: 1550.6263

Why OrientDB needs to traverse almost all records? I have tried to use desc/asc. Switch from Quiz.created_locale to Quiz.locale_created but nothings help.

Please help me with it. Thank you

**Comment by smolinari** [comment]
How many of the records have the locale "hu"? ODB probably needs to traverse all records, because the natural sort order for a date field (which I am assuming "created" is) is ascending. 

Scott

**Comment by seeden** [comment]
Hi @scot I have tried ascending sorting too but with same or similar bad results.

**Comment by smolinari** [comment]
Yeah, it doesn't look like it is possible to set an index's default sort order. 

I played with this some myself and am not sure why so many records are being read, when an index is available. This is above my head, unfortunately. Maybe @luigidellaquila can shed some more light. 

Scott

**Comment by seeden** [comment]
I have tried indexvalueasc:Quiz.locale_provider as well but with same results too

**Comment by seeden** [comment]
@lvca can you help me with it?

**Comment by seeden** [comment]
I think that this is related to #3129

**Comment by seeden** [comment]
I have tried to use type integer too. I have got same results. 
The weird thing is that
select from indexvaluesdesc:Quiz.locale_created where locale='hu' limit 20 

Returns half size:
**documentReads**: 21 603

I think that it is a bug in orientdb.
I have tried to use version 2.1.6 too with same bad results

**Comment by luigidellaquila** [comment]
Hi @seeden 

can you provide a sample dataset so that I can try it locally?

Thanks

Luigi

**Comment by seeden** [comment]
@luigidellaquila here is an OrientDB collections https://www.dropbox.com/s/2fq03duwm705th2/datatest.zip?dl=0

you can try for example query
```sql
select * from DataTest where locale='sk' order by created limit 20
```

you will see: 
**recordReads**: 5000

**Comment by seeden** [comment]
@luigidellaquila are you able to replicate this issue with my dataset? Or do you need a different format?

**Comment by seeden** [comment]
Any update ?

**Comment by StarpTech** [comment]
This issue could also effect me. What is the staus? @luigidellaquila

**Comment by seeden** [comment]
@luigidellaquila any update ?

**Comment by luigidellaquila** [comment]
Hi @seeden 

that zip contains only a small part of the db, I cannot recover from that. I'll try to reproduce the problem locally with a custom dataset.

Thanks

Luigi

**Comment by seeden** [comment]
Hi @luigidellaquila 

Let me know if you need whole db. I can provide it.
Zlatko

**Comment by luigidellaquila** [comment]
Hi @seeden 

Thank you, I managed to reproduce it locally with my db. I have a fix for it, I'm testing to make sure that it does not affect any other behaviors

Thanks

Luigi

**Comment by seeden** [comment]
You are really fast. Will be possible to use this composite index with DESC sorting too after your fix? Or is there any option how to create a DESC index for Date or Integer field (composite index)?

**Comment by luigidellaquila** [comment]
Hi @seeden 

yes, sure, there is no need to configure the index to work ASC or DESC, it can be iterated forward or backward in any case (it's true only for tree based indexes of course)

Thanks

Luigi

**Comment by seeden** [comment]
Thank you @luigidellaquila sound good. I can not wait :)

**Comment by a-unite** [comment]
+1 here!

**Comment by smolinari** [comment]
:+1: @luigidellaquila. Thanks for fixing this!

Scott

**Comment by seeden** [comment]
Hi @luigidellaquila . How is it looks like with the fix?

**Comment by luigidellaquila** [comment]
Hi @seeden 

I have the fix on a local branch, I'm just waiting for the OK from the team to merge it in 2.1.x (we are having some problems with CI tests to release the 2.1.10, the delay is because of that)

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi guys, 

this fix is now in 2.1.x branch, it will be released with 2.1.10

Thanks

Luigi

**Comment by seeden** [comment]
Thank you @luigidellaquila
